### PR TITLE
SecondLevel QueryCache fix

### PR DESCRIPTION
**Pull request by joesun99**
Hi

Please check my solution for #1385. Problem is in Loader.cs method GetResultSet where is call AutoDiscoverTypes. The resultTypes are assigned to CustomerLoader object. The class variable IType[] resultTypes is not used by reference.

Regards

### CODE REVIEW

File: src/NHibernate.Test/CacheTest/QueryCacheFixture.cs [hunk 1]
```diff
@@ -60,5 +60,19 @@ public void QueryCacheWithNullParameters()
...
 			}
 		}
+
+		[Test]
+		public void QueryCacheWithScalarFunctionThatReturnsInt()
+		{
+			using (ISession s = OpenSession())
+			{
+				int result = s
+					.CreateSQLQuery("SELECT [dbo].[TestQueryCache] ()")
```
Comment (gliljas) [inline src/NHibernate.Test/CacheTest/QueryCacheFixture.cs]: The TestQueryCache method is not in the test database. Use a simple "SELECT 1" or similar.

File: src/NHibernate/Loader/Loader.cs [hunk 2]
```diff
@@ -1533,17 +1533,17 @@ private IList ListIgnoreQueryCache(ISessionImplementor session, QueryParameters
...
 
 			QueryKey key = GenerateQueryKey(session, queryParameters);
 
-			IList result = GetResultFromQueryCache(session, queryParameters, querySpaces, resultTypes, queryCache, key);
+			IList result = GetResultFromQueryCache(session, queryParameters, querySpaces, ref resultTypes, queryCache, key);
 
 			if (result == null)
 			{
-				result = DoList(session, queryParameters, key.ResultTransformer);
+				result = DoList(session, queryParameters, queryParameters.ResultTransformer);
```
Comment (gliljas) [inline src/NHibernate/Loader/Loader.cs]: Why this change?

**Comment by hazzik** [comment]
I don't understand why types shall be passed by ref. I don't see any assignments.

File: src/NHibernate/Loader/Loader.cs (hunk 2 above)
Comment (joesun99) [reply src/NHibernate/Loader/Loader.cs]: Hi

This change is to be reverted, was inside for testing purposes.

Thx

LG

File: src/NHibernate.Test/CacheTest/QueryCacheFixture.cs (hunk 1 above)
Comment (joesun99) [reply src/NHibernate.Test/CacheTest/QueryCacheFixture.cs]: Hi

I will change test case so it will be  testable.

Regards

Comment (joesun99) [reply src/NHibernate.Test/CacheTest/QueryCacheFixture.cs]: Hi @gliljas 

I have made new feature branch https://github.com/joesun99/nhibernate-core/tree/SecondLevelQueryCacheFixOld 
Should I make new pull request and close current?

Thx

Regards

Comment (fredericDelaporte) [reply src/NHibernate.Test/CacheTest/QueryCacheFixture.cs]: Or update the current PR, unless that is two different approaches and you prefer keep the old one in case it finally should be chosen.

**Comment by joesun99** [comment]
New pull request https://github.com/nhibernate/nhibernate-core/pull/1406 in new feature branch
